src/main/java/ognl/OgnlRuntime.java

```diff
@@ -3963,19 +3966,22 @@ void put(Class clazz, String propertyName, Method method)
             if (methodsByPropertyName == null)
             {
                 methodsByPropertyName = new ConcurrentHashMap<String, Method>();
-                this.cache.put(clazz, methodsByPropertyName);
+                ConcurrentHashMap<String,Method> old = this.cache.putIfAbsent(clazz, methodsByPropertyName);
```

Comment:
`/**
     * {@inheritDoc}
     *
     * @return the previous value associated with the specified key,
     *         or {@code null} if there was no mapping for the key
     * @throws NullPointerException if the specified key or value is null
     */
    public V putIfAbsent(K key, V value) `

"if" check and then put, this is classic race condition.

Reply:
No I meant isn't following piece redundant:
```java
            ConcurrentHashMap<String,Method> methodsByPropertyName = this.cache.get(clazz);
            if (methodsByPropertyName == null)
            {
```
because you've already used `putIfAbsent` inside the `if`.

Reply:
Using read-compare-write style, which would reduce “new” operation, and very few method call new concurrent
map object.

Reply:
:clap: thanks!